docs/understand/programming_model.rst

```diff
@@ -354,12 +503,242 @@ intended use-cases.
+    for (int i = 0; i < n_streams; i++) {
+        hip_check(hipStreamCreate(&streams[i]));
+    }
+
+    // Allocate and initialize memory
+    std::vector<float> h_a(n_elements), h_b(n_elements), h_c(n_elements);
+    std::vector<float> h_ref(n_elements);  // For validation
+    for (int i = 0; i < n_elements; i++) {
+        h_a[i] = static_cast<float>(i);
+        h_b[i] = static_cast<float>(i * 2);
+        h_ref[i] = h_a[i] + h_b[i];
+    }
+
+    float *d_a, *d_b, *d_c;
+    hip_check(hipMalloc(&d_a, n_elements * sizeof(float)));
+    hip_check(hipMalloc(&d_b, n_elements * sizeof(float)));
+    hip_check(hipMalloc(&d_c, n_elements * sizeof(float)));
+
+    // Pipeline operations across streams
+    for (int i = 0; i < n_streams; i++) {
+      int offset = i * elements_per_stream;
+      int bytes = elements_per_stream * sizeof(float);
+      
+      // Stage 1: Copy input data
+      hip_check(hipMemcpyAsync(d_a + offset, h_a.data() + offset, bytes, 
+                              hipMemcpyHostToDevice, streams[i]));
+      hip_check(hipMemcpyAsync(d_b + offset, h_b.data() + offset, bytes, 
+                              hipMemcpyHostToDevice, streams[i]));
+
+      // Stage 2: Launch kernel
+      const int n_threads = 256;
+      const int n_blocks = (elements_per_stream + n_threads - 1) / n_threads;
+      vector_add<<<n_blocks, n_threads, 0, streams[i]>>>(
+          d_a + offset, d_b + offset, d_c + offset, elements_per_stream);
+      hip_check(hipGetLastError());
+
+      // Stage 3: Copy result back
+      hip_check(hipMemcpyAsync(h_c.data() + offset, d_c + offset, bytes, 
+                              hipMemcpyDeviceToHost, streams[i]));
+    }
+
+    // Wait for completion and validate
+    for (auto stream : streams) {
+      hip_check(hipStreamSynchronize(stream));
+    }
+
+    bool passed = true;
+    for (int i = 0; i < n_elements; i++) {
+      if (std::abs(h_c[i] - h_ref[i]) > 1e-5f) {
+        std::cerr << "Validation failed at " << i << ": "
+                  << h_c[i] << " != " << h_ref[i] << std::endl;
+        passed = false;
+        break;
+      }
+    }
+
+    // Cleanup
+    for (auto stream : streams) {
+      hip_check(hipStreamDestroy(stream));
+    }
+    hip_check(hipFree(d_a));
+    hip_check(hipFree(d_b));
+    hip_check(hipFree(d_c));
+
+    std::cout << "Pipeline example " << (passed ? "PASSED" : "FAILED") << std::endl;
+    return passed ? 0 : 1;
+  }
+
+There is also the producer-consumer paradigm that can be used to convert a sequential program into parallel operations to improve performance. This process can employ multiple streams to kick off asynchronous kernels, provide data to the kernels, perform operations, and return the results for further processing in the host application. 
+
+.. code_block:: cpp
+
+  #include <hip/hip_runtime.h>
+  #include <iostream>
+  #include <vector>
+  #include <cmath>
+
+  #define hip_check(hip_call)                                                   \
```

Comment:
Use the following macro which is extended with std::abort(), if you want:

https://advanced-micro-devices-demo--3739.com.readthedocs.build/projects/HIP/en/3739/how-to/hip_runtime_api/error_handling.html#hip-check-macros

Reply:
Removed example